remotes/docker/referrers.go

```diff
@@ -0,0 +1,92 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+func (r dockerFetcher) FetchReferrers(ctx context.Context, dgst digest.Digest, artifactTypes ...string) (io.ReadCloser, ocispec.Descriptor, error) {
```

Comment:
This is pretty low-level access to referrers info. Should this provide a higher level view, or be wrapped by something else? It also seems kind of weird to return a descriptor here, doesn't look like it's doing much.

- Paging
- Ensuring filtering is applied (if registry does not support filtering, or if referrers tag schema was used)
- Parsing the output to return a set of descriptors rather than an `io.ReadCloser`

Reply:
I'm thinking `artifactTypes ...string` could change to a more generic application including pages if needed. I don't think paging is a good API for referrers though and whether we should support it. Descriptor annotations are good for extra context though (paging, which endpoint, whether filters were applied, etc)

Reply:
I think not supporting paging for now is fine, we can always revisit later. Given that filtering by artifact type is not guaranteed to be supported, it would be nice if we provided client-side filtering here as a fallback.